### ISSUE TITLE
Synchronisation between wagtailuserbar and "Checks" admin panel.

### ISSUE THREAD

**Issue by ACK1D** [wagtail/wagtail#12157]
### Issue Summary

From what I see, the problems from the "Checks" tab and the Warnings data from wagtailuserbar have some kind of disynchronisation.

### Steps to Reproduce

1. Run bakerydemo with wagtail 6.2rc1
2. Login in admin
3. Go to: http://127.0.0.1:8000/recipes/mincemeat-tart/
4. You'll see the errors in the wagtailuserbar, but you won't see them in the admin panel. (Safari 17: 1 error(cross icon opens a preview, and do nothing) in admin, 14 errors in user bar)
![w_checks_1](https://github.com/user-attachments/assets/a711131a-1064-48df-a2d1-7c313bd7a816)

Having tested this on a copy of the client site, I noticed the opposite behaviour. There are no errors in wagtailuserbar, but there are errors in the admin panel. Clicking on the cross icon opens a preview, but it doesn't give any information about where the problem is. Perhaps there should be an option like debug mode (WAGTAIL_CHECKS_DEBUG = True/False) where you can see the problem in more detail for the developer or someone else. Right now these errors look like a headache for the developer from a perfectionist customer.
![w_checks_2](https://github.com/user-attachments/assets/52f4dc13-c0e5-4bbc-b8d6-2cd18d998d6d)
![w_checks_3](https://github.com/user-attachments/assets/41897a5a-b4ac-4cbf-af33-a7257c43602c)

-   I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

-   Python version: Run 3.12.2.
-   Django version: 5.0.7.
-   Wagtail version: 6.2rc1.
-   Browser version: Firefox 129.

### Working on this

Anyone can contribute to this. View our [contributing guidelines](https://docs.wagtail.org/en/latest/contributing/index.html), add a comment to the issue once you’re ready to start.

**Comment by thibaudcolas** [comment]
Hi @ACK1D, thank you for the report! I think there are multiple things going on here, each with different possible solutions. Some of those issues I don’t think are related to any changes in Wagtail 6.2 but it looks like UI changes could help alleviate them.

> You'll see the errors in the wagtailuserbar, but you won't see them in the admin panel. (Safari 17: 1 error(cross icon opens a preview, and do nothing) in admin, 14 errors in user bar)

Those specific error reports are for a color contrast accessibility check we don’t turn on by default in Wagtail because the underlying issues are unrelated to the site’s CMS. They’re also problematic because they only work correctly when the elements they’re assessing are visible. I’m guessing you had the "Checks" panel opened, rather than the Preview panel, at the point of the accessibility checker running, hence why the issues appear in one case and not the other.

This has been the case since the creation of the accessibility checker, but only with that rule enabled, so I wouldn’t recommend changes in Wagtail 6.2, but we could:

- Fix the underlying issues in the bakerydemo site
- Turn off this specific check in the demo site
- Possibly update the docs to warn against enabling the problematic rules
- In a future release, explicitly add a message in the CMS for rules we know are problematic / only appear in a specific context

> Having tested this on a copy of the client site, I noticed the opposite behaviour.

Could you share the `AccessibilityItem` customization code you’re using on that site?

There are a few possible issues here. I think they warrant investigation but likely aren’t related to the 6.2 release:

- It’s hard to understand what state of the preview page is being tested within the CMS when the preview isn’t visible. It could be that due to how your site renders, our checker goes through the page’s contents before it’s fully loaded.
- What I mentioned above about some checks running differently based on the content being visible on the screen or not.
- Possibly a related issue to #11677, where it could be that the screen resolution is different between the preview panel hidden, visible, and when you view the page full screen outside an `iframe`.

> Clicking on the cross icon opens a preview, but it doesn't give any information about where the problem is.

Of all separate parts of your report I think this one likely warrants changes in Wagtail 6.2, as it’s more directly related to the design changes we made.

- The "crosshair" button for the error reports within the page editor only takes users to the full-page preview. While the same button from the userbar highlights the specific page element where the issue is. I think we should use different icons to be clearer, or possibly implement this "element highlight" functionality within the page editor’s preview panel.
- We’ve removed the CSS selector we used to display in there from user feedback that it wasn’t particularly helpful, but I’m guessing it might be worth re-instating for the developer or power users that do find it helpful.

> Right now these errors look like a headache for the developer from a perfectionist customer.

On that specifically, I think it’s important to point out that by default we only enable a small subset of checks that have a very high probability of being related to the site’s content (as opposed to design or coding), and where editors will be able to address issues autonomously. It looks like you’ve enabled a lot more checks on your site. We’d only recommend that if:

1. You feel confident enough with the site’s accessibility that there should be few to no occasions where coding or design errors reach the live site.
2. And also you understand that if a CMS users did see any of those kinds of errors, they’d be able to report it. That’d possibly mean taking the time to override those checks’ messages in the CMS to say "report this to a developer" or similar.

**Comment by ACK1D** [comment]
> Could you share the AccessibilityItem customization code you’re using on that site?
The customization code is the same as bakery code: https://github.com/wagtail/bakerydemo/blob/676f211c0ced8b9c0b16aa248c6a03a021e4ba50/bakerydemo/base/wagtail_hooks.py#L35-L44

**Comment by thibaudcolas** [comment]
@ACK1D thanks :) Here’s the current plan for changes in Wagtail based on your feedback:

- Re-instate the previous "external destination" icon for the buttons rather than the crosshair, to make it clearer those buttons currently lead to the preview
- In the next release, implement the "highlight" of the element with the error with the preview panel. Rather than taking the user to the external preview.
- Re-instate either the CSS selector we used to display per issue, or the text of the element with the issue. TBC in what release we do this as this needs more design work.

And aside from those UI changes we’ll spend some more time reviewing how the preview rendering interacts with the checks.

**Comment by ACK1D** [comment]
@thibaudcolas thank you for your quick and detailed reply :)

**Comment by thibaudcolas** [comment]
Follow-up tasks now split so they can be worked on separately:

- #12186 (Wagtail 6.2)
- #12187 (feature request, Wagtail 6.3)
- #12188 (docs, Wagtail 6.3)
- https://github.com/wagtail/bakerydemo/issues/493
- #12189, further bug troubleshooting
